src/stream/src/executor/temporal_filter.rs

```diff
@@ -0,0 +1,479 @@
+            .zip_eq(ops.into_iter())
+            .zip_eq(lower_results.to_bitmap().iter())
+        {
+            if let Some(row_ref) = row_ref {
+                let res = lower_result;
+                match op {
+                    Op::Insert | Op::Delete => {
+                        new_ops.push(op);
+                        new_visibility.append(res);
+                    }
+                    Op::UpdateDelete => {
+                        last_res = res;
+                    }
+                    Op::UpdateInsert => {
+                        if last_res == res {
+                            new_ops.push(Op::UpdateDelete);
+                            new_ops.push(Op::UpdateInsert);
+                        } else {
+                            new_ops.push(Op::Delete);
+                            new_ops.push(Op::Insert);
+                        }
+                        new_visibility.append(last_res);
+                        new_visibility.append(res);
+                    }
+                }
+                if lower_result {
+                    let row = row_ref.to_owned_row();
+                    // Null event time should not exist in the row since it passed the filter.
+                    let timestamp = row[self.time_col_idx].as_ref().unwrap().clone();
+                    let pk = row.by_indices(self.pk_indices());
+                    match op {
+                        Op::Insert | Op::UpdateInsert => {
+                            self.buffer.insert((timestamp, pk), row.clone());
+                            self.state_table.insert(row);
+                        }
+                        Op::Delete | Op::UpdateDelete => {
+                            self.buffer.remove(&(timestamp, pk));
+                            self.state_table.delete(row);
+                        }
+                    }
+                }
+            } else {
+                new_ops.push(op);
+                new_visibility.append(false);
+            }
+        }
+
+        let new_visibility = new_visibility.finish();
+
+        (new_ops, new_visibility)
+    }
+
+    /// Fill the buffer when the temporal executor initializes or scales. On initialization
+    /// (including recovering), `prev_vnode_bitmap` will be `None`. On scaling,
+    /// `prev_vnode_bitmap` should be set as the previous vnode bitmap to perform buffer update.
+    /// Note that we do not assume set relations between `prev_vnode_bitmap` and
+    /// `curr_vnode_bitmap` when scaling occurs. That is to say, `prev_vnode_bitmap` does not
+    /// necessarily contain `curr_vnode_bitmap` on scaling out, and vice versa. Therefore, we
+    /// always check vnodes that are no longer owned and newly owned by the temporal executor
+    /// regardless of the scaling type (scale in or scale out).
+    async fn fill_buffer(
+        &mut self,
+        prev_vnode_bitmap: Option<&Bitmap>,
+        curr_vnode_bitmap: &Bitmap,
+    ) -> StreamExecutorResult<()> {
```

Comment:
> If the code snippet is to be reused, maybe we should extract it to avoid writing the same code for multiple times.

just extracted `get_rowstreams_by_new_vnodes` :-)